src/rust/engine/process_execution/src/lib.rs

```diff
@@ -275,6 +276,53 @@ impl InputDigests {
     })
   }
 
+  pub async fn new_from_merged(store: &Store, from: Vec<InputDigests>) -> Result<Self, String> {
+    let mut merged_immutable_inputs = BTreeMap::new();
+    for input_digests in from.iter() {
+      let size_before = merged_immutable_inputs.len();
+      let immutable_inputs = &input_digests.immutable_inputs;
+      merged_immutable_inputs.append(&mut immutable_inputs.clone());
+      if size_before + immutable_inputs.len() != merged_immutable_inputs.len() {
+        return Err(
+          "Tried to merge two-or-more immutable inputs at the same path with different values!"
+            .to_string(),
```

Comment:
Would be good to render a bit more explanation... maybe:
```suggestion
          format!("Tried to merge two-or-more immutable inputs at the same path with different values! The collision involved one of the entries in: {immutable_inputs:?}"),
```
(untested)